### ISSUE TITLE
fallback to grey framed theme for older browsers

### ISSUE THREAD

**Issue by gnadelwartz** [webmin/webmin#859]
authentic does no more support older browsers, e.g. ie11. from time to time there are complaints about this. the solution, evenb don't liked us use grey framed theme.

I can understand they want to use authentic advanced features, but also must use ie11 from company.
see https://github.com/qooob/authentic-theme/issues/818 for some rants.

a possible solution may be to check if the theme supports this browser, if not fallback to grey framed.

this can be done similar to ask modules, if a theme has a file theme_check.pl it's included and the function theme_check ust called with the browser string as first argument. if false is returned webmin can fall back to grey framed.

**Comment by qooob** [comment]
It's important to check first if Virtualmin Framed Theme is installed and choose it, otherwise Gray Theme.

Thanks, Kay.

**Comment by jcameron** [comment]
This sounds good in principal, but the hard part will be determining which browser user-agent strings aren't supported. They are notoriously hard to parse due to a complete lack of standardization!

**Comment by gnadelwartz** [comment]
yes, this is why I propose to do it on theme side as qooob is brilliant annthis 😏

and in addtion I would propose to check for well known incompatible main stream browsers.

**Comment by qooob** [comment]
Not sure what you're talking about, but it's done the same way on server-side just as on client-side.

```Perl
$ENV{'HTTP_USER_AGENT'} =~ /MSIE/
```

**Comment by gnadelwartz** [comment]
that's what I say it should be easy to detect msie, and all versions are incompatible, same with lynx and w3mir

http://www.useragentstring.com/pages/useragentstring.php?name=Lynx
http://www.useragentstring.com/pages/useragentstring.php?name=w3m

**Comment by swelljoe** [comment]
I didn't notice that IE11 had been dropped completely. That's problematic. It still has a significant market share (as much as 9%, depending on who you believe).

Is it not possible to use some sort of polyfill for IE11? I'm fine with dropping support for all earlier IE versions, but 11 still has some market share.

**Comment by gnadelwartz** [comment]
https://github.com/paulmillr/es6-shim

**Comment by qooob** [comment]
The first, or the second line in changelog for 19.00 was the message about dropping all IE support, around six months ago. There are many  serious undealeable/very hard to deal with issues, one of them is [redirects](https://github.com/yiisoft/yii2/issues/9670) but there are many more (which now I don't remember about already). Simply using shims is not going to solve the problem easily.

I strongly recommend using server-side to detect IE and load the old theme, with the message on _System Information_ page and in _Change Language and Theme_, and _Webmin Configuration/Language_.

I'm fine with using _babeljs_ to convert ES6 code for older browsers to work, like, `stats.js`, and by the way, mail in the very near future, but supporting IE is over-kill.

If I did that, it would slow down things tremendously.

**Comment by gnadelwartz** [comment]
also the last one who complains is fine with auto detection and fallback, the complaints are more about that some customers use ie11, so the fallback to a workung theme for them and stay with authentic for all others is a step in the right direction

**Comment by jcameron** [comment]
There is some precedent for doing this server-side in Webmin already - mobile devices are detected this way, and can use a custom theme automatically. 

What happens with IE11 in the authentic theme currently? Does it at least display a message to the user about compatability?

**Comment by qooob** [comment]
Yes, it does display the nice message, explaining the situation.

**Comment by gnadelwartz** [comment]
yes, there 8s a (annoying) popup warning about ie11 is no more supported in every mouse klick

**Comment by qooob** [comment]
It's not annoying, it's protective! ☺️ I tend not to come up with anything annoying.

**Comment by gnadelwartz** [comment]
@jcameron are you considering implementing something like this? if yes we should consider an interims release because of cron issue and fallback implementation for ie11 users.

**Comment by jcameron** [comment]
I'll likely do a 1.881 release once the cron issue is sorted out. However, this fallback will take longer to implement as it changes the theme selection behavior significantly..

**Comment by qooob** [comment]
I think it should and could be done on the fly, without chanign any config files.

Once this kind of force mode is activated, the user will not be able to choose Authentic Theme (option in select will be disabled) and near the fields on pages where you could change the theme will be written in parentheses, possibly with red text, that it's the compatibility mode. If we factor out the function and the theme could definitely it, than any theme could decide which browser should run in compatibility mode. I think for this time being good thing to have.

I'm planning to finish 19.10 with tree view in File Manager before the end of the week. There were too many other request, that distracted me. I wish 19.10 go to 1.881.

**Comment by gnadelwartz** [comment]
If admin has selected authentic as default (and want it tgis way) this will not help, because a regular user is already on authentic-theme.

Disabling setting and display a warning may an help to prevent regular users to complaint why they can't use authenetic, even selected.

**Comment by qooob** [comment]
Force to Gray/VMF Theme for all browsers, listed in new subroutine (that theme could override) as deprecated. _System Information_ page will have dismissable message. _Webmin Configuration/Language_ and _Change Language and Theme_ will have the text in parentheses that  say compatibility mode. All of this will be the result of new sub, for example `sub theme_unsupported`, so in case subroutine returns true and given theme set as default, the rules above are applied.

**Comment by gnadelwartz** [comment]
an other question regarding fallback:

if a module like e.g. Commodo WAF is installed and known to be incompatible with new authentic-theme AND there is no easy fix. Should this also reason to trigger a fallback?

**Comment by qooob** [comment]
Yes, great idea. So the subroutine could also accept moduels names. Not sure how easy it's gonna be to make it fall back for the single module only.

**Comment by gnadelwartz** [comment]
I think it's all or nothing, but not on module level

**Comment by qooob** [comment]
All or nothing? So once user installs old module that doesn't work in the recent version of the theme the whole installation should switch to the old theme? No, no. It should be selectively for modules and all for browser's.

**Comment by gnadelwartz** [comment]
changing theme on the fly per module would be hard, lets see what Jamie think's.

an alternative may be to open incompatile modules in a sperate window from theme or webmin.

**Comment by jcameron** [comment]
Per-module doesn't seem viable, since the authentic theme changes the way modules are rendered on the page. Also there exists both framed and non-framed themes that can't be mixed in a single session.

A per-user setting could work though. In fact, a work-around that you could use now is to create an extra Webmin user with a different theme set, and login as that user on IE11.

**Comment by qooob** [comment]
Extra user doesn't seem to be an option for those guys.

I personally think, all of this is very strange, that you can't simply download a portable version of Firefox, and run it without installing.

On the theme side there is a user friendly message about no support for Internet Explorer.

The only valuable improvement that I see here, is to make server push old theme when unsupported browser is used. 

For example, when I use Gnome's Epiphany web-browser with Google Drive, I get a message, that using this browser might cause issues. It's absolutely normal practice apparently. Even giants can't support all possible browsers and make things just work; probably they could but that would slow development a lot.

I will update the message, proposing to use portable version of Firefox.

**Comment by wolfbeast** [comment]
> I personally think, all of this is very strange, that you can't simply download a portable version of Firefox, and run it without installing.

As indicated above, often this is the case of corporate policies, where "downloading and running" anything third party that isn't company-prescribed is also a big no-no. Same for separate users; in these cases user administration is usually completely centralized so a user is actually a user/person in the company, and different login credentials are not allowed.

**Comment by qooob** [comment]
@jcameron Jamie, is this possible to add it on your side? It would be useful for the users?

**Comment by jcameron** [comment]
Yes, this could be implemented ... I just need to find time to work on it.

As a work-around, I'd recommend using a separate root-equivalent user for now (and for the user for whom  you want to use the new theme, have them select it explicitly in the Change Language and Theme module, and then set the global default to the older theme).

**Comment by dimangelid** [comment]
Is there a way to disable the warning about the unsupported browser?

**Comment by qooob** [comment]
Technically it's possible but it will not work anyway.
